tools/dotnet-linker/Steps/ManagedRegistrarStep.cs

```diff
@@ -0,0 +1,1162 @@
+		// emit garbage and really weird things start happening at runtime.
+		bool? trace;
+		void Trace (ILProcessor il, string message)
+		{
+			if (!trace.HasValue)
+				trace = !string.IsNullOrEmpty (Environment.GetEnvironmentVariable ("XAMARIN_MSR_TRACE"));
+			if (trace.Value) {
+				il.Emit (OpCodes.Ldstr, message);
+				il.Emit (OpCodes.Call, abr.Runtime_TraceCaller);
+			}
+		}
+
+		int counter;
+		void CreateUnmanagedCallersMethod (MethodDefinition method, AssemblyTrampolineInfo infos)
+		{
+			var baseMethod = StaticRegistrar.GetBaseMethodInTypeHierarchy (method);
+			var placeholderType = abr.System_IntPtr;
+			ParameterDefinition? callSuperParameter = null;
+			VariableDefinition? returnVariable = null;
+			var leaveTryInstructions = new List<Instruction> ();
+			var isVoid = method.ReturnType.Is ("System", "Void");
+
+			var name = $"callback_{counter++}_{Sanitize (method.DeclaringType.FullName)}_{Sanitize (method.Name)}";
+
+			var callbackType = method.DeclaringType.NestedTypes.SingleOrDefault (v => v.Name == "__Registrar_Callbacks__");
+			if (callbackType is null) {
+				callbackType = new TypeDefinition (string.Empty, "__Registrar_Callbacks__", TypeAttributes.NestedPrivate | TypeAttributes.Sealed | TypeAttributes.Class);
+				callbackType.BaseType = abr.System_Object;
+				method.DeclaringType.NestedTypes.Add (callbackType);
+			}
+
+			var callback = callbackType.AddMethod (name, MethodAttributes.Public | MethodAttributes.Static | MethodAttributes.HideBySig, placeholderType);
+			callback.CustomAttributes.Add (CreateUnmanagedCallersAttribute (name));
+			infos.Add (new TrampolineInfo (callback, method, name));
+
+			// If the target method is marked, then we must mark the trampoline as well.
+			method.CustomAttributes.Add (CreateDynamicDependencyAttribute (callbackType, callback.Name));
+
+			var body = callback.CreateBody (out var il);
+			var placeholderInstruction = il.Create (OpCodes.Nop);
+			var placeholderNextInstruction = il.Create (OpCodes.Nop);
+			var postProcessing = new List<Instruction> ();
+			var categoryAttribute = StaticRegistrar.GetCategoryAttribute (method.DeclaringType);
+			var isCategory = categoryAttribute is not null;
+			var isInstanceCategory = isCategory && StaticRegistrar.HasThisAttribute (method);
+			var isGeneric = method.DeclaringType.HasGenericParameters;
+			VariableDefinition? selfVariable = null;
+
+			Trace (il, $"ENTER");
+
+			if (method.IsConstructor) {
+				callback.AddParameter ("pobj", abr.ObjCRuntime_NativeHandle);
+			} else {
+				callback.AddParameter ("pobj", abr.System_IntPtr);
+			}
+
+			if (!isVoid || method.IsConstructor)
+				returnVariable = body.AddVariable (placeholderType);
+
+			if (isGeneric) {
+				if (method.IsStatic)
+					throw ErrorHelper.CreateError (4130 /* The registrar cannot export static methods in generic classes ('{0}'). */, method.FullName);
+
+				il.Emit (OpCodes.Ldtoken, method);
+
+				il.Emit (OpCodes.Ldarg_0);
```

Comment:
```suggestion
				il.Emit (OpCodes.Ldarg_0);
				if (method.IsConstructor) {
					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
				}
```

This is minimal change necessary to fix https://github.com/xamarin/xamarin-macios/pull/18268#issuecomment-1545895343.  Obviously it would be nicer to just pass `NativeHandle` to the correct overload but that would be a slightly larger change.

Reply:
The fix looks reasonable. I'll do extra validation run on the sample w/ NativeAOT.

Reply:
Confirmed to be fixed.